lib/MusicBrainz/Server/Data/Role/QueryToList.pm

```diff
@@ -16,13 +16,19 @@ sub query_to_list {
 }
 
 sub query_to_list_limited {
-    my ($self, $query, $args, $limit, $offset, $builder) = @_;
+    my ($self, $query, $args, $limit, $offset, $builder, $dollar_placeholders) = @_;
 
     $builder //= $self->can('_new_from_row');
 
     my @args = @$args;
+    my $count = @args;
+
     if (defined $offset) {
-        $query .= " OFFSET ?";
+        if ($dollar_placeholders) {
+            $query .= ' OFFSET $' . (++$count);
```

Comment:
> Initially $count is size of @args, right? So you increment it to reference $offset and $limit below?

Right, it's to match the position of offset and limit in the parameter list.

> Also, what's the difference between using ? placeholders and $<value>? I don't understand why there need to be two ways to pass these arguments.

The `$` placeholders let you reference a parameter multiple times in the query, by its position. With `?`, you have to pass the same parameter multiple times, in the positions they appear.

A problem is, you can't mix these styles in the same query, or it'll throw an error. Since `query_to_list_limited` is taking an outside query and wrapping it with extra placeholders, it needs to know what style the inner query is using.

The specific reason I used `$` placeholders is, the query construction in `subscribed_entity_edits` is really complicated. When I had to worry about pushing parameters in the correct order, it was even worse.